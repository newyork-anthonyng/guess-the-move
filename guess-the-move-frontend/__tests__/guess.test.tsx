import Guess from "../src/app/guess/page";
import { render, screen, fireEvent } from "@testing-library/react";
import userEvent from "@testing-library/user-event";
import "@testing-library/jest-dom";
import { expect } from "@jest/globals";
import { Chessground } from "chessground";
import { useRouter } from "next/navigation";

let config;
let mockGetFen = jest.fn();
let mockSet = jest.fn(a => config = { ...config, ...a });

jest.mock("chessground", () => ({
  Chessground: jest.fn(($div, _config) => {
    config = _config;

    return {
      getFen: mockGetFen,
      set: mockSet
    }
  })
}));

let mockRouter;
jest.mock("next/navigation", () => ({
  useRouter: jest.fn(),
}));

beforeEach(() => {
  mockRouter = {
    push: jest.fn()
  }
  useRouter.mockReturnValue(mockRouter);
});

it('should render error if game is invalid', async () => {
  render(<Guess searchParams={{ gameId: 'invalid' }} />);

  await screen.findByText(/game does not exist/i);
});

it.only("should set up chessground correctly", () => {
  render(<Guess />);

  expect(Chessground).toHaveBeenCalledTimes(1);
  expect(Chessground.mock.calls[0][1]).toMatchSnapshot();
});

it("should show evaluation after playing a move", async () => {
  render(<Guess />);

  // play e4
  mockGetFen.mockImplementationOnce(() => 'rnbqkbnr/pppppppp/8/8/4P3/8/PPPP1PPP/RNBQKBNR');
  config.events.move();

  await screen.findByText(/your move's evaluation is:/i);
  await screen.findByText('-1');
  await screen.findByText(/in the game, they played/i);
  await screen.findByText('+1');

  let $backButton = screen.getByRole('button', { name: /back/i });
  await userEvent.click($backButton);

  expect(mockSet).toHaveBeenCalledTimes(1);
  expect(mockSet.mock.calls[0][0]).toMatchSnapshot();

  expect(screen.queryByText(/your move's evaluation is:/i)).toBeNull();
  expect(screen.queryByText(/in the game, they played/i)).toBeNull();
  await screen.findByText(/play the best move/i);

  expect(mockSet).toHaveBeenCalledTimes(2);
  expect(mockSet.mock.calls[1][0]).toMatchSnapshot();

  // play Nf3
  mockGetFen.mockImplementationOnce(() => 'rnbqkbnr/pppp1ppp/8/4p3/4P3/5N2/PPPP1PPP/RNBQKB1R');
  config.events.move();

  await screen.findByText(/your move's evaluation is:/i);
  await screen.findByText('-2');
  await screen.findByText(/in the game, they played/i);
  await screen.findByText('+2');

  $backButton = screen.getByRole('button', { name: /back/i });
  await userEvent.click($backButton);

  expect(mockSet).toHaveBeenCalledTimes(3);
  expect(mockSet.mock.calls[2][0]).toMatchSnapshot();

  expect(screen.queryByText(/your move's evaluation is:/i)).toBeNull();
  expect(screen.queryByText(/in the game, they played/i)).toBeNull();
  await screen.findByText(/play the best move/i);

  expect(mockSet).toHaveBeenCalledTimes(4);
  expect(mockSet.mock.calls[3][0]).toMatchSnapshot();

  // play Bc4
  mockGetFen.mockImplementationOnce(() => 'r1bqkbnr/pppp1ppp/2n5/4p2Q/2B1P3/8/PPPP1PPP/RNB1K1NR');
  config.events.move();

  await screen.findByText(/your move's evaluation is:/i);
  await screen.findByText('-3');
  await screen.findByText(/in the game, they played/i);
  await screen.findByText('+3');

  $backButton = screen.getByRole('button', { name: /back/i });
  await userEvent.click($backButton);

  expect(mockSet).toHaveBeenCalledTimes(5);
  expect(mockSet.mock.calls[4][0]).toMatchSnapshot();

  expect(screen.queryByText(/your move's evaluation is:/i)).toBeNull();
  expect(screen.queryByText(/in the game, they played/i)).toBeNull();
  await screen.findByText(/play the best move/i);

  expect(mockSet).toHaveBeenCalledTimes(6);
  expect(mockSet.mock.calls[5][0]).toMatchSnapshot();

  // play Qxf7#
  mockGetFen.mockImplementationOnce(() => 'r1bqkbnr/pppp1Qp1/2n4p/4p3/2B1P3/8/PPPP1PPP/RNB1K1NR');
  config.events.move();

  await screen.findByText(/your move's evaluation is:/i);
  await screen.findByText('-4');
  await screen.findByText(/in the game, they played/i);
  await screen.findByText('+4');

  $backButton = screen.getByRole('button', { name: /back/i });
  await userEvent.click($backButton);

  const $finishButton = screen.getByRole('button', { name: /finish/i });
  await userEvent.click($finishButton);

  expect(mockRouter.push).toHaveBeenCalledTimes(1); 
  expect(mockRouter.push.mock.calls[0][0]).toEqual('/results'); 
});
