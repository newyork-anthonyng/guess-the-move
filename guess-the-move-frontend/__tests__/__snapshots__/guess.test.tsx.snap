// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should set up chessground correctly 1`] = `
{
  "events": {
    "move": [Function],
  },
  "movable": {
    "dests": Map {
      "a2" => [
        "a3",
        "a4",
      ],
      "b2" => [
        "b3",
        "b4",
      ],
      "c2" => [
        "c3",
        "c4",
      ],
      "d2" => [
        "d3",
        "d4",
      ],
      "e2" => [
        "e3",
        "e4",
      ],
      "f2" => [
        "f3",
        "f4",
      ],
      "g2" => [
        "g3",
        "g4",
      ],
      "h2" => [
        "h3",
        "h4",
      ],
      "b1" => [
        "a3",
        "c3",
      ],
      "g1" => [
        "f3",
        "h3",
      ],
    },
    "free": false,
    "showDests": true,
  },
}
`;

exports[`should show evaluation after playing a move 1`] = `
{
  "comments": [
    "",
  ],
  "fen": "rnbqkbnr/pppppppp/8/8/4P3/8/PPPP1PPP/RNBQKBNR b KQkq - 0 1",
  "lastMove": [
    "e2",
    "e4",
  ],
  "movable": {
    "dests": Map {
      "b8" => [
        "c6",
        "a6",
      ],
      "g8" => [
        "h6",
        "f6",
      ],
      "a7" => [
        "a6",
        "a5",
      ],
      "b7" => [
        "b6",
        "b5",
      ],
      "c7" => [
        "c6",
        "c5",
      ],
      "d7" => [
        "d6",
        "d5",
      ],
      "e7" => [
        "e6",
        "e5",
      ],
      "f7" => [
        "f6",
        "f5",
      ],
      "g7" => [
        "g6",
        "g5",
      ],
      "h7" => [
        "h6",
        "h5",
      ],
    },
    "free": false,
    "showDests": true,
  },
  "moveId": "/?",
  "san": "e4",
  "startingComments": undefined,
  "turnColor": "black",
  "uci": "e2e4",
}
`;

exports[`should show evaluation after playing a move 2`] = `
{
  "comments": [
    "",
  ],
  "fen": "rnbqkbnr/pppp1ppp/8/4p3/4P3/8/PPPP1PPP/RNBQKBNR w KQkq - 0 2",
  "lastMove": [
    "e7",
    "e5",
  ],
  "movable": {
    "dests": Map {
      "a2" => [
        "a3",
        "a4",
      ],
      "b2" => [
        "b3",
        "b4",
      ],
      "c2" => [
        "c3",
        "c4",
      ],
      "d2" => [
        "d3",
        "d4",
      ],
      "f2" => [
        "f3",
        "f4",
      ],
      "g2" => [
        "g3",
        "g4",
      ],
      "h2" => [
        "h3",
        "h4",
      ],
      "b1" => [
        "a3",
        "c3",
      ],
      "d1" => [
        "e2",
        "f3",
        "g4",
        "h5",
      ],
      "e1" => [
        "e2",
      ],
      "f1" => [
        "e2",
        "d3",
        "c4",
        "b5",
        "a6",
      ],
      "g1" => [
        "e2",
        "f3",
        "h3",
      ],
    },
    "free": false,
    "showDests": true,
  },
  "moveId": "WG",
  "san": "e5",
  "startingComments": undefined,
  "turnColor": "white",
  "uci": "e7e5",
}
`;

exports[`should show evaluation after playing a move 3`] = `
{
  "comments": [
    "",
  ],
  "fen": "rnbqkbnr/pppp1ppp/8/4p2Q/4P3/8/PPPP1PPP/RNB1KBNR b KQkq - 1 2",
  "lastMove": [
    "d1",
    "h5",
  ],
  "movable": {
    "dests": Map {
      "b8" => [
        "c6",
        "a6",
      ],
      "d8" => [
        "e7",
        "f6",
        "g5",
        "h4",
      ],
      "e8" => [
        "e7",
      ],
      "f8" => [
        "e7",
        "d6",
        "c5",
        "b4",
        "a3",
      ],
      "g8" => [
        "h6",
        "f6",
        "e7",
      ],
      "a7" => [
        "a6",
        "a5",
      ],
      "b7" => [
        "b6",
        "b5",
      ],
      "c7" => [
        "c6",
        "c5",
      ],
      "d7" => [
        "d6",
        "d5",
      ],
      "g7" => [
        "g6",
        "g5",
      ],
      "h7" => [
        "h6",
      ],
    },
    "free": false,
    "showDests": true,
  },
  "moveId": "&J",
  "san": "Qh5",
  "startingComments": undefined,
  "turnColor": "black",
  "uci": "d1h5",
}
`;

exports[`should show evaluation after playing a move 4`] = `
{
  "comments": [
    "",
  ],
  "fen": "r1bqkbnr/pppp1ppp/2n5/4p2Q/4P3/8/PPPP1PPP/RNB1KBNR w KQkq - 2 3",
  "lastMove": [
    "b8",
    "c6",
  ],
  "movable": {
    "dests": Map {
      "h5" => [
        "g6",
        "f7",
        "h6",
        "h7",
        "h4",
        "h3",
        "g4",
        "f3",
        "e2",
        "d1",
        "g5",
        "f5",
        "e5",
      ],
      "a2" => [
        "a3",
        "a4",
      ],
      "b2" => [
        "b3",
        "b4",
      ],
      "c2" => [
        "c3",
        "c4",
      ],
      "d2" => [
        "d3",
        "d4",
      ],
      "f2" => [
        "f3",
        "f4",
      ],
      "g2" => [
        "g3",
        "g4",
      ],
      "h2" => [
        "h3",
        "h4",
      ],
      "b1" => [
        "a3",
        "c3",
      ],
      "e1" => [
        "e2",
        "d1",
      ],
      "f1" => [
        "e2",
        "d3",
        "c4",
        "b5",
        "a6",
      ],
      "g1" => [
        "e2",
        "f3",
        "h3",
      ],
    },
    "free": false,
    "showDests": true,
  },
  "moveId": "\\M",
  "san": "Nc6",
  "startingComments": undefined,
  "turnColor": "white",
  "uci": "b8c6",
}
`;

exports[`should show evaluation after playing a move 5`] = `
{
  "comments": [
    "",
  ],
  "fen": "r1bqkbnr/pppp1ppp/2n5/4p2Q/2B1P3/8/PPPP1PPP/RNB1K1NR b KQkq - 3 3",
  "lastMove": [
    "f1",
    "c4",
  ],
  "movable": {
    "dests": Map {
      "a8" => [
        "b8",
      ],
      "d8" => [
        "e7",
        "f6",
        "g5",
        "h4",
      ],
      "e8" => [
        "e7",
      ],
      "f8" => [
        "e7",
        "d6",
        "c5",
        "b4",
        "a3",
      ],
      "g8" => [
        "h6",
        "f6",
        "e7",
      ],
      "a7" => [
        "a6",
        "a5",
      ],
      "b7" => [
        "b6",
        "b5",
      ],
      "d7" => [
        "d6",
        "d5",
      ],
      "g7" => [
        "g6",
        "g5",
      ],
      "h7" => [
        "h6",
      ],
      "c6" => [
        "b8",
        "e7",
        "d4",
        "b4",
        "a5",
      ],
    },
    "free": false,
    "showDests": true,
  },
  "moveId": "(=",
  "san": "Bc4",
  "startingComments": undefined,
  "turnColor": "black",
  "uci": "f1c4",
}
`;

exports[`should show evaluation after playing a move 6`] = `
{
  "comments": [
    "",
  ],
  "fen": "r1bqkbnr/pppp1pp1/2n4p/4p2Q/2B1P3/8/PPPP1PPP/RNB1K1NR w KQkq - 0 4",
  "lastMove": [
    "h7",
    "h6",
  ],
  "movable": {
    "dests": Map {
      "h5" => [
        "g6",
        "f7",
        "h6",
        "h4",
        "h3",
        "g4",
        "f3",
        "e2",
        "d1",
        "g5",
        "f5",
        "e5",
      ],
      "c4" => [
        "b5",
        "a6",
        "d5",
        "e6",
        "f7",
        "d3",
        "e2",
        "f1",
        "b3",
      ],
      "a2" => [
        "a3",
        "a4",
      ],
      "b2" => [
        "b3",
        "b4",
      ],
      "c2" => [
        "c3",
      ],
      "d2" => [
        "d3",
        "d4",
      ],
      "f2" => [
        "f3",
        "f4",
      ],
      "g2" => [
        "g3",
        "g4",
      ],
      "h2" => [
        "h3",
        "h4",
      ],
      "b1" => [
        "a3",
        "c3",
      ],
      "e1" => [
        "e2",
        "f1",
        "d1",
      ],
      "g1" => [
        "e2",
        "f3",
        "h3",
      ],
    },
    "free": false,
    "showDests": true,
  },
  "moveId": "ZR",
  "san": "h6",
  "startingComments": undefined,
  "turnColor": "white",
  "uci": "h7h6",
}
`;
